src/main/scala/midas/passes/PreLinkRenaming.scala

```diff
@@ -0,0 +1,51 @@
+// See LICENSE for license details.
+
+package midas.passes
+
+import firrtl._
+import firrtl.annotations.{CircuitName, ModuleName}
+import firrtl.ir._
+import firrtl.Mappers._
+import Utils._
+import java.io.{File, FileWriter, StringWriter}
+
+// This transform renames a circuit's modules and annotations in
+// preparation for linking (eg. PlatformMapping). Subsumes Utils.renameMods
+private[passes] class PreLinkRenaming(parentNamespace: Namespace) extends firrtl.Transform {
+
+  override def name = "[MIDAS] Pre-link Module Renaming"
+  // TODO: Technically this xform should be able to accept AnyForm(TM); and be form idempotent
+  def inputForm = LowForm
+  def outputForm = LowForm
+
+  // Updates instantiations of modules that would alias
+  def updateInsts(nameMap: Map[String, String])(s: Statement): Statement = s match {
+    case s: WDefInstance if nameMap.contains(s.module) => s.copy(module = nameMap(s.module))
+    case s => s.map(updateInsts(nameMap))
+  }
+  // Renames the module if there is a collision, and updates submodule inst names
+  def updateModNames(nameMap: Map[String, String])(m: DefModule): DefModule = (m match {
+      case m : ExtModule => m.copy(name = nameMap.getOrElse(m.name, m.name))
+      case m :    Module => m.copy(name = nameMap.getOrElse(m.name, m.name))
+  }) map updateInsts(nameMap)
+
+  def execute(state: CircuitState): CircuitState = {
+    val circuit = state.circuit
+    require(!parentNamespace.contains(circuit.main), "Submodule in child has same name as top")
+
+    // Find all modules that share a name with a module in the circuit we are linking against
+    val modsToRename = circuit.modules.collect({
```

Comment:
Could just be `circuit.modules.map(m => m.name -> parentNamespace.newName(m.name))`. Aside from being simpler, this makes sure that the parent namespace gets the names of all the child modules rather than just the ones that are renamed. The namespace might not be used for anything later, but it's nice to keep it consistent.

This also lets you drop the `contains` checks elsewhere.